src/seedu/addressbook/data/person/ContactDetail.java

```diff
@@ -0,0 +1,84 @@
+package seedu.addressbook.data.person;
+
+import seedu.addressbook.data.exception.IllegalValueException;
+
+/**
+ * Represents a person's contact detail
+ * Guarantees: immutable; is valid as declared in {@link #isValid(String)}
+ */
+public abstract class ContactDetail {
+    /**
+     * 
+     */
+    private String value;
+
+    /**
+     * Instance's value attribute will be set to String supplied to constructor
+     * @param value
+     * @throws IllegalValueException Thrown if value is invalid, error message implemented by inherited class
+     */
+    public ContactDetail(String value) throws IllegalValueException {
+        this.value = value.trim();
+        if (!isValid(value)) {
+            throw new IllegalValueException(getValidationErrorMessage());
+        }
+    }
+
+    public void printClass() {
+        System.out.print(this.getClass());
+    }
+
+    public static void main(String[] args) {
+        try {
+            new Email("valid@e.mail", false).printClass();
+            ;
+        } catch (IllegalValueException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Returns error message if validation fails
+     * @return
+     */
+
+    abstract String getValidationErrorMessage();
+
+    /**
+     * Returns true if a given string is a valid value
+     */
+    public boolean isValid(String test) {
+        return test.matches(getValidationRegex());
+    }
+
+    /**
+     * Returns regex used for validation
+     * @return
+     */
+    abstract String getValidationRegex();
+
+    /**
+     * @return the value
+     */
+    public String getValue() {
+        return value;
+    }
+
+    @Override
+    public int hashCode() {
+        return value.hashCode();
+    }
+
+    @Override
+    public String toString() {
+        return value;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other.getClass().equals(this.getClass()) // instanceof handles nulls
```

Comment:
According to [Java 8 API](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#getClass()) getClass() returns the runtime class of the Object. It wouldn't make sense if there is equality between two objects of different classes which both inherit from a superclass class because in Java all classes inherit Object. Whereas according to Oracle's documentation the conditional operator [instanceof](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/op2.html) can be used to "test if an object is an instance of a class, an instance of a subclass, or an instance of a class that implements a particular interface"